File: go/day01.go
package main

import (
	"fmt"
	"sort"
	"strconv"
)

type ElfFood struct {
	num      int
	calories int
}

type Day01 struct {
	elf         int
	elfCalories int
	elves       []ElfFood
}

func (self *Day01) Parse(line string) {
	if line == "" {
		fmt.Printf("Elf n. %d holding %d calories\n", self.elf, self.elfCalories)
		self.elves = append(self.elves, ElfFood{self.elf, self.elfCalories})
		self.elf++
		self.elfCalories = 0
	} else {
		cal, err := strconv.Atoi(line)
		if err != nil {
			fmt.Printf("Invalid line: %s - %v\n", line, err)
		}
		self.elfCalories += cal
	}
}

func (self *Day01) Solve() {
	sort.Slice(self.elves, func(i, j int) bool {
		return self.elves[i].calories > self.elves[j].calories
	})
	fmt.Printf("Highest calories: %d\n", self.elves[0].calories)
	var top3 = 0
	for i := 0; i < 3; i++ {
		top3 += self.elves[i].calories
	}
	fmt.Printf("Top 3 calories: %d\n", top3)
}

File: src/day01.py
from advent import Solver


# https://adventofcode.com/2022/day/1
# quite simple, of course just go through it in one pass of the file.
# both solutions can be calculated while reading the file
class Solution(Solver):
    def __init__(self):
        self.elf = 0
        self.elf_calories = 0
        self.calories = []

    def file_name(self):
        return "../files/day01-calories.txt"

    def test_data(self):
        return """1000
2000
3000

4000

5000
6000

7000
8000
9000

10000"""

    def parse(self, line: str):
        if not line:
            # print(f"Elf n. {self.elf} holding {self.elf_calories} calories")
            self.calories.append([self.elf, self.elf_calories])
            self.elf_calories = 0
            self.elf += 1
            return
        self.elf_calories += int(line)

    def solve(self):
        self.calories.append([self.elf, self.elf_calories])
        self.calories.sort(key=lambda x: x[1], reverse=True)
        print("Saw {} elves: maximum: {}".format(self.elf, self.calories[0]))
        top3 = [x[1] for x in self.calories[0:3]]
        print("First 3 elves: {}".format(sum(top3)))
